File: zoomimage/src/main/java/mcgars/com/zoomimage/ZoomImageController.java
package mcgars.com.zoomimage;

import android.app.Activity;
import android.os.Build;
import android.support.v4.view.ViewPager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import com.alexvasilkov.gestures.animation.ViewPositionAnimator;

import java.util.List;

import mcgars.com.zoomimage.adapter.ZoomPhotoPagerAdapter;
import mcgars.com.zoomimage.ui.Displayer;
import mcgars.com.zoomimage.utils.DecorUtils;

/**
 * Created by Владимир on 23.03.2016.
 * Зумирует картинку на передний план
 */
public class ZoomImageController implements ViewPositionAnimator.PositionUpdateListener {
    private Activity activity;
    private Displayer displayer;
    private ViewPager zoomPager;
    private ViewGroup root;
    private View back;
    private Toolbar toolbar;
    private ZoomPhotoPagerAdapter flicAdapter;
    private View vShadow;

    public ZoomImageController(Activity activity, Displayer displayer){
        this.activity = activity;
        this.displayer = displayer;
        init();
    }

    private void init() {

        if(zoomPager !=null)
            return;

        ViewGroup decorView = (ViewGroup) activity.findViewById(android.R.id.content);
        if (decorView != null) {
            root = decorView;
            zoomPager = (ViewPager) root.findViewById(R.id.zoomFullPager);
            if(zoomPager !=null) {
                initViews();
                return;
            } else {
                LayoutInflater inflater = LayoutInflater.from(activity);
                View v = inflater.inflate(R.layout.zoom_view_image, root, false);
                root.addView(v);

                zoomPager = (ViewPager) v.findViewById(R.id.zoomFullPager);
            }
            initViews();
        }

        // full image toolbar
        toolbar.setNavigationIcon(R.drawable.ic_arrow_back_black_24dp);
        toolbar.setNavigationOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                activity.onBackPressed();
            }
        });

        initDecorMargins();
    }

    private void initViews() {
        back = root.findViewById(R.id.flickrFullBackground);
        toolbar = (Toolbar) root.findViewById(R.id.flickrFullToolbar);
        vShadow = root.findViewById(R.id.vShadow);
    }

    public Toolbar getZoomToolbar() {
        return toolbar;
    }

    public View getBack() {
        return back;
    }

    public ViewPager getZoomPager() {
        return zoomPager;
    }

    public void initFlickrAdapter(){
        if(flicAdapter == null) {
            flicAdapter = new ZoomPhotoPagerAdapter(zoomPager, displayer);
            flicAdapter.setPositionAnimationListener(this);
        }
    }

    @Override
    public void onPositionUpdate(float state, boolean isLeaving) {
        back.setVisibility(state == 0f ? View.INVISIBLE : View.VISIBLE);
        back.getBackground().setAlpha((int) (255 * state));

        vShadow.getBackground().setAlpha((int) (255 * state));
        vShadow.setVisibility(state == 0f ? View.INVISIBLE : View.VISIBLE);

        toolbar.setVisibility(state == 0f ? View.INVISIBLE : View.VISIBLE);
        if (Build.VERSION.SDK_INT > 10)
            toolbar.setAlpha(state);

        Log.d(getClass().getName(), "onPositionUpdate: " + state);

        if (isLeaving && state == 0f) {
            flicAdapter.setActivated(false);
        }
    }

    public void onDestroy() {
        if(flicAdapter!=null)
            flicAdapter.removePositionAnimation(this);
    }

    public boolean onBackPressed() {
        if (flicAdapter != null && !flicAdapter.getAnimator().isLeaving()) {
            flicAdapter.getAnimator().exit(true);
            return true;
        }
        return false;
    }

    public void setPhotos(List<? extends ZoomPhotoPagerAdapter.IPhoto> data) {
        zoomPager.setAdapter(flicAdapter);
        flicAdapter.setPhotos(data);
    }

    public ZoomImageController setPhotos(ImageView from, List<? extends ZoomPhotoPagerAdapter.IPhoto> data) {
        initFlickrAdapter();
        flicAdapter.from(from);
        setPhotos(data);
        return this;
    }

    public ZoomImageController setPhotos(List<ImageView> from, List<? extends ZoomPhotoPagerAdapter.IPhoto> data) {
        initFlickrAdapter();
        flicAdapter.from(from);
        setPhotos(data);
        return this;
    }

    public ZoomImageController setPhotos(ViewPager from, List<? extends ZoomPhotoPagerAdapter.IPhoto> data) {
        initFlickrAdapter();
        flicAdapter.from(from);
        setPhotos(data);
        return this;
    }

    public ZoomImageController setPhotos(RecyclerView from, List<? extends ZoomPhotoPagerAdapter.IPhoto> data) {
        initFlickrAdapter();
        flicAdapter.from(from);
        setPhotos(data);
        return this;
    }

    public void enter(int position) {
        flicAdapter.setActivated(true);
        flicAdapter.getAnimator().enter(position, true);
    }

    protected void initDecorMargins() {
        // Adjusting margins and paddings to fit translucent decor
        if(Build.VERSION.SDK_INT >= 16 ) {
        if(root.getChildAt(0).getFitsSystemWindows())
            DecorUtils.marginForStatusBar(toolbar);
        }
    }
}

File: zoomimage/src/main/java/mcgars/com/zoomimage/ZoomImageController.kt
package mcgars.com.zoomimage

import androidx.viewpager.widget.ViewPager
import com.alexvasilkov.gestures.animation.ViewPositionAnimator
import mcgars.com.zoomimage.adapter.ZoomPhotoPagerAdapter
import mcgars.com.zoomimage.fabric.AMBuilder
import mcgars.com.zoomimage.model.IPhoto
import mcgars.com.zoomimage.ui.Displayer
import mcgars.com.zoomimage.ui.UiContainer

class ZoomImageController(
        private val uiContainer: UiContainer,
        private val displayer: Displayer,
        private val titleTransformer: ((Int, Int, IPhoto?) -> CharSequence)? = null
) : ViewPositionAnimator.PositionUpdateListener, ViewPager.OnPageChangeListener {

    private val zoomAdapter: ZoomPhotoPagerAdapter by lazy {
        ZoomPhotoPagerAdapter(uiContainer.zoomPager, displayer, this)
    }

    init {
        uiContainer.zoomPager.addOnPageChangeListener(this)
        uiContainer.zoomToolbar.setNavigationOnClickListener { onBackPressed() }
    }

    /**
     * Animate cosmetics views (ViewPager, Toolbar, Background)
     * @param state float
     * @param isLeaving boolean
     */
    override fun onPositionUpdate(state: Float, isLeaving: Boolean) {
        uiContainer.show(invisible = state == 0f, alpha = (255 * state).toInt())

        if (isLeaving && state == 0f) {
            zoomAdapter.setActivated(false)
        }
    }

    fun onDestroy() {
        uiContainer.zoomPager.removeOnPageChangeListener(this)
        zoomAdapter.removePositionAnimation()
    }

    /**
     * If image is full size closed it
     * @return if image is open true else false
     */
    fun onBackPressed(): Boolean {
        return zoomAdapter.exit(true)
    }

    /**
     * Before [.show] need set data and builder
     *
     * @param builder set from views
     * @param data images
     * @return this
     */
    fun setPhotos(builder: AMBuilder, data: List<IPhoto>): ZoomImageController {
        zoomAdapter.from(builder)
        setPhotos(data)
        return this
    }

    private fun setPhotos(data: List<IPhoto>) {
        uiContainer.zoomPager.adapter = zoomAdapter
        zoomAdapter.setPhotos(data)
    }

    /**
     * Zoom image to front
     *
     * @param position by
     */
    fun show(position: Int = 0) {
        zoomAdapter.setActivated(true)
        zoomAdapter.enter(position, true)
        // fixed, if firs page onPageSelected not trigger
        if (position == 0) setTitle(position)
    }

    override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {

    }

    override fun onPageSelected(position: Int) {
        setTitle(position)
    }

    private fun setTitle(position: Int) {
        val item = zoomAdapter.getPhoto(position)

        val title = titleTransformer?.invoke(position, zoomAdapter.count, item) ?: when {
            zoomAdapter.count <= 1 -> item?.text
            else -> createDefaultTitle(position, item)
        }

        uiContainer.zoomToolbar.title = title
    }

    private fun createDefaultTitle(position: Int, item: IPhoto?): String {
        val text = (position + 1).toString() + " / " + zoomAdapter.count
        return when {
            item?.text.isNullOrEmpty().not() -> "(" + text + ") " + item?.text
            else -> text
        }
    }

    override fun onPageScrollStateChanged(state: Int) {

    }
}
